ESPS/general/man/man3/minrelent.3
.\" Copyright (c) 1988-1990 Entropic Speech, Inc.
.\" Copyright (c) 1997 Entropic Research Laboratory, Inc. All rights reserved.
.\" @(#)minrelent.3	1.2 18 Apr 1997 ESI/ERL
.ds ]W (c) 1997 Entropic Research Laboratory, Inc.
.TH MIN_REL_ENT 3\-ESPSsp 18 Apr 1997
.if t .ds S \(*S
.if n .ds S \fRSUM\fP
.if t .ds f \fI
.if n .ds f \fR
.if t .ds a \(*a
.if n .ds a "alpha 
.if t .ds b \(*b
.if n .ds b beta
.SH NAME
min_rel_ent \- compute maximum-entropy or minmum-relative-entropy estimate of probability distribution
.SH SYNOPSIS
.ft B
.nf
double
min_rel_ent(p, c, q, beta, m, n, maxit, thresh)
    double  *p, **c, *q, *beta;
    int	    m, n, maxit;
    double  thresh;
.SH DESCRIPTION
.PP
Given
.I p,
an
.I initial estimate
of a probability distribution,
and
.I c,
a
.I constraint matrix,
the function
.I min_rel_ent
computes
.I q,
a
.I final estimate
of the probatility distribution.
This is a vector with nonnegative elements
\*fq\d\s-3j\s+3\u\fR
that satisfies the
.I constraints
.IP
\*f\*S\d\s-3j\s+3\u c\d\s-3ij\s+3\uq\d\s-3j\s+3\u\fR = 0,
.IP
\*f\*S\d\s-3j\s+3\u q\d\s-3j\s+3\u\fR = 1,
.LP
and, subject to those constraints, minimizes the
.I relative entropy,
.IP
\*f\*S\d\s-3j\s+3\u q\d\s-3j\s+3\u \fRlog\fP q\d\s-3j\s+3\u/p\d\s-3j\s+3\u\fR
.LP
of
.I q
with respect to
.I p.
The solution is of the form
.IP
\*fq\d\s-3j\s+3\u = \*ap\d\s-3j\s+3\u \fRexp\fP \*S\d\s-3i\s+3\u \*b\d\s-3i\s+3\uc\d\s-3ij\s+3\u\fR
.LP
where \*a and the elements of \*b are Lagrange multipliers chosen to
satisfy the constraints.  The value of \*b is available as an additional
output
.I beta.
.PP
To obtain a maximum-entropy estimate
.I q
subject to the stated constraints,
use an initial estimate
.I p
whose elements are all equal.
.PP
The function argument
.I p
should be an array of
.I n
nonnegative numbers
(or more precisely a pointer to the first element of such an array).
The sum of the elements of
.I p
need not be normalized to 1, and the same results should be obtained
as if each element were divided by the sum.
The argument
.I c
should be an
.I m
by
.I n
matrix, represented as an array of
.I m
pointers to rows of length
.I n.
(Such structures can be allocated by
.IR arr_alloc (3-ESPSu)).
The arguments
.I q
and
.I beta
should be arrays of lengths
.I n
and
.I m
to hold the results;
these must be allocated by the calling program.
The arguments
.I m
and
.I n
give the array dimensions.
The argument
.I thresh
is a convergence criterion.
The function uses an iterative procedure that terminates as soon as
the relative change in the computed value of every element of
.I q
is less than
.I thresh
from one iteration to the next.
If the convergence criterion is not satisfied after
.I maxit
iterations, the procedure prints an error message and terminates anyway.
The function value returned by
.I min_rel_ent
is the maximum relative change in any element of
.I q
between the last iteration and the next to last.
It is less than
.I thresh
in case of successful termination
and not otherwise.
.SH DIAGNOSTICS
Function value greater than or equal to
.I thresh;
error message:
.IP
min_rel_ent: convergence failed after \fIn\fP iterations.
.SH BUGS
None known.
.SH REFERENCES
.LP
1.  R. Johnson,
``Determining Probability Distributions by Maximum Entropy
and Minimum Cross-Entropy,''
.I APL Quote Quad,
vol. 9, no. 4, June 1979, pp. 24-29.
(APL 79 Conference Proceedings).
.LP
2.  J. Shore and R. Johnson,
``Axiomatic Derivation of the Principle of Maximum Entropy
and the Principle of Minimum Relative Entropy,''
.I IEEE Trans. Information Theory,
vol. IT-26, no. 1, pp. 26-37, Jan. 1980
.SH SEE ALSO
.nf
\fIrel_ent\fP(3\-ESPSsp)
.fi
.SH AUTHOR
Rodney W. Johnson, Entropic Speech, Inc.
Based on an APL function from Ref. 2
and on a Fortran adataption by Joseph T. Buck.


ESPS/general/src/lib_sp/minrelent.c
/*----------------------------------------------------------------------+
|									|
|   This material contains proprietary software of Entropic Speech,	|
|   Inc.  Any reproduction, distribution, or publication without the	|
|   prior written permission of Entropic Speech, Inc. is strictly	|
|   prohibited.  Any public distribution of copies of this work		|
|   authorized in writing by Entropic Speech, Inc. must bear the	|
|   notice								|
|									|
|    "Copyright (c) 1988 Entropic Speech, Inc. All rights reserved."	|
|									|
+-----------------------------------------------------------------------+
|									|
|  min_rel_ent.c							|
|									|
|  Rodney W. Johnson.  Based on an APL function from the paper		|
|	referenced below and on a Fortran adataption by Joseph T.	|
|	Buck.								|
|									|
|  min_rel_ent() computes a minimum-relatve-entropy probability		|
|	distribution q, given an initial estimate p and a constraint	|
|	matrix c.  The result is a vector of nonnegative numbers that	|
|	minimizes the relative entropy of q with respect to p,		|
|	subject to the constraints that the matrix product cp equal	|
|	zero and the sum of the elements of q be 1.  The solution is	|
|	of the form							|
|	                      m-1					|
|	    q  = alpha p  exp SUM beta  c     (j = 0, . . . , n-1)	|
|	     j          j     i=0     i  ij				|
|	where alpha and the elements of beta are Lagrange multipliers	|
|	chosen to satisfy the constraints.  The value of beta is	|
|	available as an additional output.				|
|									|
|   Ref:  R. Johnson, "Determining Probability Distributions by		|
|	Maximum Entropy and Minimum Cross-Entropy", APL Quote Quad,	|
|	vol. 9, no. 4, June 1979, pp. 24-29.  (APL 79 Conference	|
|	Proceedings).							|
|									|
+----------------------------------------------------------------------*/
#ifndef lint
    static char *sccs_id = "@(#)minrelent.c	1.1	9/2/89	ESI";
#endif

#include <stdio.h>
#include <esps/esps.h>
#include <math.h>

char	*arr_alloc();
void	arr_free();

static double	dist();
int		lsq_solv();


double
min_rel_ent(p, c, q, beta, m, n, maxit, thresh)
    double  *p, **c, *q, *beta;
    int	    m, n, maxit;
    double  thresh;
{
    double  *rp, *q0;
    double  **t;
    double  *delta_b;
    double  sum, term, d;
    long    dim[2];
    int	    i, j, iter;

    dim[0] = n;
    rp = (double *) arr_alloc(1, dim, DOUBLE, 0);
    q0 = (double *) arr_alloc(1, dim, DOUBLE, 0);
    dim[1] = m;
    t = (double **) arr_alloc(2, dim, DOUBLE, 0);
    dim[0] = m;
    delta_b = (double *) arr_alloc(1, dim, DOUBLE, 0);

    for (j = 0; j < n; j++)
	q[j] = rp[j] = sqrt(p[j]);

    for (i = 0; i < m; i++)
	beta[i] = 0;

    for (   iter = 0,  d = DBL_MAX;
	    iter < maxit && d >= thresh;
	    iter++,    d = dist(q, q0, n) )
    {
	for (j = 0; j < n; j++)
	    q0[j] = q[j];

	for (i = 0; i < m; i++)
	    for (j = 0; j < n; j++)
		t[j][i] = c[i][j]*q[j];

	(void) lsq_solv(t, q, n, m, delta_b, thresh);

	for (i = 0; i < m; i++)
	    beta[i] -= delta_b[i];

	for (j = 0; j < n; j++)
	{
	    term = 0;

	    for (i = 0; i < m; i++)
		term += beta[i]*c[i][j];

	    q[j] = rp[j] * exp(0.5*term);
	}
    }

    if (d > thresh)
	Fprintf(stderr,
	    "min_rel_ent: convergence failed after %d iterations.\n",
	    maxit);

    sum = 0.0;

    for (j = 0; j < n; j++)
	sum += q[j] *= q[j];

    for (j = 0; j < n; j++)
	q[j] /= sum;

    arr_free((char *) rp, 1, DOUBLE, 0);
    arr_free((char *) q0, 1, DOUBLE, 0);
    arr_free((char *) t, 2, DOUBLE, 0);
    arr_free((char *) delta_b, 1, DOUBLE, 0);

    return d;
}

static double
dist(p, q, n)
    double  *p, *q;
    int	    n;
{
    int	    j;
    double  a, b, c;
    double  d = 0.0;

    for (j = 0; j < n; j++)
    {
	c = p[j] - q[j];
	if (c != 0.0)
	{
	    a = fabs(p[j]);
	    b = fabs(q[j]);
	    c = (a > b) ? fabs(c)/a : fabs(c)/b;
	    if (c > d) d = c;
	}
    }

    return d;
}
